src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,274 @@
+
+@[ext] theorem dilation.ext {f g : dilation α β} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `dilation` with a new `to_fun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def dilation.copy (f : dilation α β) (f' : α → β) (h : f' = ⇑f) : dilation α β :=
+{ to_fun := f',
+  edist_eq' := h.symm ▸ f.edist_eq' }
+
+/-- The ratio of a dilation `f`. Uses `Exists.some` -/
+def ratio {α β} [pseudo_emetric_space α] [pseudo_emetric_space β] {F : Type*}
+  [dilation_class F α β] (f : F) : ℝ≥0 :=
+(dilation_class.edist_eq' f).some
+
+end setup
+
+namespace dilation
+
+variables {α : Type*} {β : Type*} {γ : Type*} {F : Type*} {G : Type*}
+
+@[simp] lemma edist_eq
+  [pseudo_emetric_space α] [pseudo_emetric_space β] [dilation_class F α β]
+  (f : F) (x y : α) : edist (f x) (f y) = ratio f * edist x y :=
+(dilation_class.edist_eq' f).some_spec x y
+
+@[simp] lemma dist_eq
+  [pseudo_metric_space α] [pseudo_metric_space β] [dilation_class F α β]
+  (f : F) (x y : α) : dist (f x) (f y) = ratio f * dist x y :=
+by simp only [dist_edist, edist_eq, ennreal.to_real_mul, ennreal.coe_to_real]
+
+theorem nndist_eq [pseudo_metric_space α] [pseudo_metric_space β] [dilation_class F α β]
+  (f : F) (x y : α) : nndist (f x) (f y) = ratio f * nndist x y :=
+begin
+  apply subtype.ext _,
+  simp only [coe_nndist, dist_eq, nonneg.coe_mul],
+end
+
+section pseudo_emetric_dilation
+
+variables [pseudo_emetric_space α] [pseudo_emetric_space β] [pseudo_emetric_space γ]
+variables [dilation_class F α β] [dilation_class G β γ]
+variables (f : F) (g : G) {x y z : α}  {s : set α}
+
+lemma lipschitz : lipschitz_with (ratio f) (f : α → β) :=
+λ x y, (edist_eq f x y).le
+
+
+-- TODO: add `instance ennreal.div_inv_comm_monoid`
+-- TODO: fix `antilipschitz_with` decl header
+lemma antilipschitz (hr : ratio f ≠ 0) : antilipschitz_with ((ratio f)⁻¹) (f : α → β) :=
+λ x y, begin
+  -- `div_eq_mul_inv` should be `div_eq_inv_mul`
+  rw [mul_comm, ← ennreal.div_le_iff_le_mul, div_eq_mul_inv, mul_comm, ennreal.coe_inv hr],
+  simp only [inv_inv, edist_eq, le_refl],
+  left, simp [hr],
+  left, simp [hr],
+end
+
+/-- A dilation from an emetric space is injective -/
+lemma injective {α : Type*} [emetric_space α]  [dilation_class F α β] (f : F) (hr : ratio f ≠ 0) :
+  injective f := (dilation.antilipschitz f hr).injective
+
+/-- Any map on a subsingleton is a dilation -/
+def of_subsingleton [subsingleton α] (f : α → β) : dilation α β :=
+{ to_fun := f,
+  edist_eq' := ⟨0, λ x y, by { rw subsingleton.elim x y, simp, }⟩ }
+
+/-- The composition of similarities is a dilation -/
+def comp (f : dilation α β) (g : dilation β γ):
+  dilation α γ :=
+{ to_fun := g ∘ f,
+  edist_eq' := ⟨ratio f * ratio g, λ x y, by { simp only [edist_eq, ennreal.coe_mul], ring, }⟩ }
+
+/-- The constant function of is a dilation -/
+def const (b : β) :
+  dilation α β :=
+{ to_fun := λ _, b,
+  edist_eq' := ⟨0, λ x y, by simp⟩ }
```

Comment:
In addition to these two definitions, there should be one for the identify map as a dilation. All three definitions should then have various trivial lemmas (e.g. the result of applying the resulting dilations at points or coercing them to functions, the results of composing with the identity or constant map on either side, associativity of composition, the `ratio` of these maps in all cases where it's uniquely defined such as for a `nontrivial` `metric_space`), mostly marked as `simp` lemmas.